File: app/src/main/java/com/example/myshop/adapter/ProductAdapter.java
package com.example.myshop.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.myshop.R;
import com.example.myshop.domain.Category;
import com.example.myshop.domain.Product;

import java.util.ArrayList;

public class ProductAdapter extends RecyclerView.Adapter<ProductAdapter.ViewHolder> {
    ArrayList<Product>categoryProduct;

    public ProductAdapter(ArrayList<Product> categoryProduct) {

        this.categoryProduct = categoryProduct;
    }

    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View inflate= LayoutInflater.from(parent.getContext()).inflate(R.layout.viewholder_category,parent, false);
        return new ViewHolder(inflate);
    }

    @Override
    public void onBindViewHolder(@NonNull ProductAdapter.ViewHolder holder, int position) {
        holder.categoryName.setText(categoryProduct.get(position).getTitle());
        String picUrl = "";
        switch (position){
            case 0:{
                picUrl = "womens";
                holder.mainLayout.setBackground(ContextCompat.getDrawable(holder.itemView.getContext(),R.drawable.cat_background));
                break;
            }
            case 1: {
                picUrl = "mens";
                holder.mainLayout.setBackground(ContextCompat.getDrawable(holder.itemView.getContext(), R.drawable.cat_background));
                break;
            }
            case 2: {
                picUrl = "jewelery";
                holder.mainLayout.setBackground(ContextCompat.getDrawable(holder.itemView.getContext(), R.drawable.cat_background));
                break;
            }
            case 3: {
                picUrl = "electronics";
                holder.mainLayout.setBackground(ContextCompat.getDrawable(holder.itemView.getContext(), R.drawable.cat_background));
                break;
            }
        }
        int drawableResourceId = holder.itemView.getContext().getResources().getIdentifier(picUrl, "drawable", holder.itemView.getContext().getPackageName());

        Glide.with(holder.itemView.getContext())
                .load(drawableResourceId)
                .into(holder.categoryPic);
    }

    @Override
    public int getItemCount() {

        return categoryProduct.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder{
        TextView title, fee;
        ImageView pic;
        TextView addBtn;

        public ViewHolder(@NonNull View itemView){
            super(itemView);
            title=itemView.findViewById(R.id.title);
            fee=itemView.findViewById(R.id.fee);
            pic=itemView.findViewById(R.id.pic);
            addBtn=itemView.findViewById(R.id.addBtn);

        }
    }
}

File: app/src/main/java/com/example/my_shopping/ProductAdapter.kt
package com.example.my_shopping

import android.app.Activity
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide

private fun ItemProduct.count() {

}

private operator fun Any.compareTo(i: Int): Int {
    TODO("Not yet implemented")
}

private operator fun Any.inc() {

}

private operator fun Any.dec() {

}

class ProductAdapter(private val Products: ArrayList<ItemProduct>) : RecyclerView.Adapter<ProductAdapter.ViewHolder>() {

    var onClickItem: ((ItemProduct) -> Unit)? = null

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val title: TextView = itemView.findViewById(R.id.product_title)
        val category: TextView = itemView.findViewById(R.id.category)
        val price: TextView = itemView.findViewById(R.id.price)
        val image: ImageView = itemView.findViewById(R.id.product_Image)
        val countButton: TextView = itemView.findViewById(R.id.countButton)
        val plusButton: ImageView = itemView.findViewById(R.id.plusButton)
        val minusButton: ImageView = itemView.findViewById(R.id.minusButton)
        val removeButton: TextView = itemView.findViewById(R.id.removeButton)
        val mainLayout: ConstraintLayout = itemView.findViewById(R.id.item_layout)

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view =
            LayoutInflater.from(parent.context).inflate(R.layout.information_product, parent, false)
        return ViewHolder(view)
    }

    override fun getItemCount(): Int {
        return Products.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val product = Products[position]
        holder.title.text = product.name
        holder.category.text = product.category
        holder.price.text = product.price.toString() + "$"

        if (product.count() > 0) {
            holder.countButton.visibility = View.VISIBLE
            holder.plusButton.visibility = View.VISIBLE
            holder.minusButton.visibility = View.VISIBLE
            holder.removeButton.visibility = View.VISIBLE
            holder.countButton.text = product.count().toString()
        } else if (product.count() <= 0) {
            holder.countButton.visibility = View.INVISIBLE
            holder.plusButton.visibility = View.INVISIBLE
            holder.minusButton.visibility = View.INVISIBLE
            holder.removeButton.visibility = View.INVISIBLE
            holder.countButton.text = product.count().toString()
        }

        val url = product.image
        Glide.with(holder.itemView.context)
            .load(url)
            .into(holder.image)

        holder.plusButton.setOnClickListener {
            holder.countButton.text = product.count().toString()

            val context = holder.itemView.context
            if (context is Activity) {
                context.finish()
                context.overridePendingTransition(0, 0)
            }
            val intent = Intent(context, Cart::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        }

        holder.minusButton.setOnClickListener {

            if (product.count() <= 0) {
                holder.countButton.visibility = View.INVISIBLE
                holder.plusButton.visibility = View.INVISIBLE
                holder.minusButton.visibility = View.INVISIBLE
                holder.removeButton.visibility = View.INVISIBLE

                val context = holder.itemView.context
                if (context is Activity) {
                    context.finish()
                    context.overridePendingTransition(0, 0)
                }
                val intent = Intent(context, Cart::class.java)
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(intent)

            }
            holder.countButton.text = product.count().toString()

            val context = holder.itemView.context
            if (context is Activity) {
                context.finish()
                context.overridePendingTransition(0, 0)
            }
            val intent = Intent(context, Cart::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        }
        holder.removeButton.setOnClickListener() {
            holder.countButton.visibility = View.INVISIBLE
            holder.plusButton.visibility = View.INVISIBLE
            holder.minusButton.visibility = View.INVISIBLE
            holder.removeButton.visibility = View.INVISIBLE
            holder.countButton.text = product.count().toString()

            val context = holder.itemView.context
            if (context is Activity) {
                context.finish()
                context.overridePendingTransition(0, 0)
            }
            val intent = Intent(context, Cart::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        }
        holder.itemView.setOnClickListener {
            onClickItem?.invoke(product)
        }
    }
}


